src/reconstruction/community.jl

```diff
@@ -0,0 +1,406 @@
+            ),
+        )
+    end
+
+    # get offsets to construct community S
+    reaction_lengths = [n_reactions(model) for model in models]
+    metabolite_lengths = [n_metabolites(model) for model in models]
+    reaction_offset = [0; cumsum(reaction_lengths[1:end-1])]
+    metabolite_offset = [0; cumsum(metabolite_lengths[1:end-1])]
+
+    # get each model's S matrix (needed for the size calculations)
+    stoichs = [stoichiometry(model) for model in models]
+    nnzs = [findnz(stoich) for stoich in stoichs] # nonzero indices per model
+
+    # size calculations
+    column_add = add_biomass_objective ? 1 : 0 # objective rxn
+    row_add = add_biomass_objective ? length(models) : 0 # biomass as metabolites
+    nnz_add = add_biomass_objective ? (1 + length(models)) : 0
+    nnz_total =
+        sum(length(first(nnz)) for nnz in nnzs) +
+        length(models) * length(exchange_rxn_ids) +
+        length(exchange_met_ids) + nnz_add
+    n_reactions_metabolic = sum(reaction_lengths)
+    n_reactions_total = n_reactions_metabolic + length(exchange_rxn_ids) + column_add
+    n_metabolites_metabolic = sum(metabolite_lengths)
+    n_metabolites_total = n_metabolites_metabolic + length(exchange_met_ids) + row_add
+
+    # Create empty storage vectors
+    lbs = spzeros(n_reactions_total)
+    ubs = spzeros(n_reactions_total)
+    rxns = Array{String,1}(undef, n_reactions_total)
+    mets = Array{String,1}(undef, n_metabolites_total)
+    I = Array{Int,1}(undef, nnz_total)
+    J = Array{Int,1}(undef, nnz_total)
+    V = Array{Float64,1}(undef, nnz_total)
+
+    # build metabolic components, block diagonals
+    kstart = 1
+    for i = 1:length(models)
+        kend = kstart + length(nnzs[i][3]) - 1
+        rng = kstart:kend
+        I[rng] .= nnzs[i][1] .+ metabolite_offset[i]
+        J[rng] .= nnzs[i][2] .+ reaction_offset[i]
+        V[rng] .= nnzs[i][3]
+        kstart += length(nnzs[i][3])
+    end
+
+    # build environmental - exchange links
+    for i = 1:length(models)
+        exchange_rxn_inds = indexin(exchange_rxn_ids, reactions(models[i]))
+        exchange_met_inds = indexin(exchange_met_ids, metabolites(models[i]))
+        for (n, (ex_rxn, ex_met)) in enumerate(zip(exchange_rxn_inds, exchange_met_inds)) # each exchange rxn has one exchange met
+            isnothing(ex_rxn) && continue
+            isnothing(ex_met) && continue
+            # connect environmental metabolite with exchange metabolite
+            I[kstart] = n_metabolites_metabolic + n
+            J[kstart] = ex_rxn + reaction_offset[i]
+            V[kstart] = -stoichs[i][ex_met, ex_rxn] # ex is normally negative, make positive
+            kstart += 1
+        end
+    end
+
+    # # build diagonal environmental exchanges
+    for i = 1:length(exchange_rxn_ids)
+        I[kstart] = n_metabolites_metabolic + i
+        J[kstart] = n_reactions_metabolic + i
+        V[kstart] = -1.0
+        kstart += 1
+    end
+
+    if add_biomass_objective
```

Comment:
shouldn't it be called `set_biomass_objective`? it's technically not _adding_ anything.

Reply:
actually it does get added to matrix here